include/libtorrent/storage.hpp

```diff
@@ -194,6 +194,18 @@ namespace aux {
 		mutable std::mutex m_file_created_mutex;
 		mutable typed_bitfield<file_index_t> m_file_created;
 
+#if TORRENT_HAVE_MAP_VIEW_OF_FILE
+		// Windows has a race condition when unmapping a view while a new
+		// view or mapping object is being created in a different thread.
+		// The race can cause a page of written data to be zeroed out before
+		// it is written out to disk. To avoid the race these calls must be
+		// serialized on a per-file basis. See github issue #3842 for details.
+
+		// This array stores a mutex for each file in the storage object
+		// It must be aquired before calling CreateFileMapping or UnmapViewOfFile
+		mutable std::shared_ptr<std::mutex[]> m_file_open_unmap_lock;
```

Comment:
storing arrays in `shared_ptr` is a c++17 feature. afaik, master currently builds with C++14. Do you think we should just go straight to 17?

Reply:
I believe the use of a custom deleter is enough to make this work with C++14, is it not?

Reply:
I think you may have to hide that it's an array too, like `std::shared_ptr<std::mutex>`, with the custom deleter knowing that it's an array.